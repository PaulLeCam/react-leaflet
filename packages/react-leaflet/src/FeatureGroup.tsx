import {
  createElementObject,
  createPathComponent,
  extendContext,
  type PathProps,
} from '@react-leaflet/core'
import { FeatureGroup as LeafletFeatureGroup } from 'leaflet'

import type { LayerGroupProps } from './LayerGroup.js'

export interface FeatureGroupProps extends LayerGroupProps, PathProps {}

export const FeatureGroup = createPathComponent<
  LeafletFeatureGroup,
  FeatureGroupProps
>(function createFeatureGroup({ children: _c, ...options }, ctx) {
  const group = new LeafletFeatureGroup([], options)
  return createElementObject(
    group,
    extendContext(ctx, { layerContainer: group, overlayContainer: group }),
  )
})
