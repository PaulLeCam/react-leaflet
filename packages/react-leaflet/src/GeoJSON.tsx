import {
  createElementObject,
  createPathComponent,
  extendContext,
  type PathProps,
} from '@react-leaflet/core'
import type { GeoJsonObject } from 'geojson'
import { type GeoJSONOptions, GeoJSON as LeafletGeoJSON } from 'leaflet'

import type { LayerGroupProps } from './LayerGroup.js'

export interface GeoJSONProps
  extends GeoJSONOptions,
    LayerGroupProps,
    PathProps {
  data: GeoJsonObject
}

export const GeoJSON = createPathComponent<LeafletGeoJSON, GeoJSONProps>(
  function createGeoJSON({ data, ...options }, ctx) {
    const geoJSON = new LeafletGeoJSON(data, options)
    return createElementObject(
      geoJSON,
      extendContext(ctx, { overlayContainer: geoJSON }),
    )
  },
  function updateGeoJSON(layer, props, prevProps) {
    if (props.style !== prevProps.style) {
      if (props.style == null) {
        layer.resetStyle()
      } else {
        layer.setStyle(props.style)
      }
    }
  },
)
