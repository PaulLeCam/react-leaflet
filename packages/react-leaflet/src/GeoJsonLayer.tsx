import {
  Feature,
  GeoJSON,
  LineString,
  MultiLineString,
  MultiPoint,
  MultiPolygon,
  Point,
  Polygon as PolygonType,
  Position,
} from 'geojson'
import { GeoJSON as LeafletGeoJSON, LatLng, LeafletEventHandlerFnMap } from 'leaflet'
import React from 'react'
import {
  CircleMarker,
  CircleMarkerProps,
  FeatureGroup,
  GeoJSONProps,
  Marker,
  MarkerProps,
  Polygon,
  PolygonProps,
  Polyline,
  PolylineProps,
} from 'react-leaflet'

import { hasProperty, isGeometry } from './typeGuards'


export type GeoJsonLayerProps = GeoJSONProps & {
  data: GeoJSON
  pointToLayer?: never
  onEachFeature?: never
  coordsToLatLng?: (coords: number[]) => LatLng

  feature?: Feature
  eventHandlerFn?: (feature: Feature) => LeafletEventHandlerFnMap | undefined
  getChildren?: (feature: Feature) => React.ReactNode
  lineProps?: Partial<PolylineProps> | (
    (featureLine?: Feature<LineString | MultiLineString>) => Partial<PolylineProps>
  )
  polygonProps?: Partial<PolygonProps> | (
    (featurePolygon?: Feature<PolygonType | MultiPolygon>) => Partial<PolylineProps>
  )
} & ({
  pointAsMarker: true
  pointProps?: Partial<MarkerProps> | (
    (featurePoint?: Feature<Point | MultiPoint>) => Partial<MarkerProps>
  )
} | {
  pointAsMarker?: false
  pointProps?: Partial<CircleMarkerProps> | (
    (featurePoint?: Feature<Point | MultiPoint>) => Partial<CircleMarkerProps>
  )
})

const defaultFilter = () => true

const defaultCoordsToLatLng = (coords: Position) => LeafletGeoJSON.coordsToLatLng([
  coords[0] ?? NaN,
  coords[1] ?? NaN,
  coords[2] ?? NaN,
])

const GeoJsonLayer: React.FC<GeoJsonLayerProps> = ({
  data,
  style,
  filter = defaultFilter,
  coordsToLatLng = defaultCoordsToLatLng,

  eventHandlerFn,
  feature,
  pointAsMarker,
  pointProps,
  lineProps,
  polygonProps,
  getChildren,

  children,
  ...featureGroupProps
}) => {
  // Override (N.B.: only for feature group)
  featureGroupProps.pathOptions = {
    ...featureGroupProps.pathOptions,
    ...(typeof style === 'function' ? style(isGeometry(feature?.geometry) ? feature : undefined) : style)
  }

  switch (data.type) {

    case 'Point':  // Fall-through
    case 'MultiPoint': {
      const points = data.type === 'Point' ? [data.coordinates] : data.coordinates
      const markerProps = typeof pointProps === 'function'
        ? pointProps(feature as Feature<Point | MultiPoint> | undefined)
        : pointProps
      if (pointAsMarker) {
        return (
          <FeatureGroup {...featureGroupProps}>
            {points.map((position, i) => (
              <Marker {...markerProps} key={i} position={coordsToLatLng(position)} />
            ))}
            {children}
          </FeatureGroup>
        )
      }
      return (
        <FeatureGroup {...featureGroupProps}>
          {points.map((position, i) => (
            <CircleMarker {...markerProps} key={i} center={coordsToLatLng(position)} />
          ))}
          {children}
        </FeatureGroup>
      )
    }

    case 'LineString': // Fall-through
    case 'MultiLineString': {
      const lines = data.type === 'LineString' ? [data.coordinates] : data.coordinates
      const polylineProps = typeof lineProps === 'function'
        ? lineProps(feature as Feature<LineString | MultiLineString> | undefined)
        : lineProps
      return (
        <FeatureGroup {...featureGroupProps}>
          {lines.map((line, i) => (
            <Polyline {...polylineProps} key={i} positions={line.map(position => coordsToLatLng(position))} />
          ))}
          {children}
        </FeatureGroup>
      )
    }

    case 'Polygon': // Fall-through
    case 'MultiPolygon': {
      const polygons = data.type === 'Polygon' ? [data.coordinates] : data.coordinates
      const polyProps = typeof polygonProps === 'function'
        ? polygonProps(feature as Feature<PolygonType | MultiPolygon> | undefined)
        : polygonProps
      return (
        <FeatureGroup {...featureGroupProps}>
          {polygons.map((polygon, i) => (
            <Polygon
              {...polyProps}
              key={i}
              positions={polygon.map(line => line.map(position => coordsToLatLng(position)))}
            />
          ))}
          {children}
        </FeatureGroup>
      )
    }

    case 'Feature': // Fall-through
    case 'FeatureCollection': {
      const features = data.type === 'Feature' ? [data] : data.features
      return (
        <FeatureGroup {...featureGroupProps}>
          {features.filter(filter).map((feature, i) => (
            <GeoJsonLayer
              {...featureGroupProps}

              key={feature.id ?? (
                hasProperty(feature.properties, 'id', 'string', 'number')
                  ? feature.properties.id
                  : i
              )}

              data={feature.geometry}
              style={style}
              filter={filter}
              coordsToLatLng={coordsToLatLng}
              eventHandlers={eventHandlerFn ? eventHandlerFn(feature) : undefined}
              children={getChildren ? getChildren(feature) : undefined}

              eventHandlerFn={eventHandlerFn}
              getChildren={getChildren}
              feature={feature}
              pointAsMarker={pointAsMarker}
              pointProps={pointProps}
              lineProps={lineProps}
              polygonProps={polygonProps}
            />
          ))}
          {children}
        </FeatureGroup>
      )
    }

    case 'GeometryCollection': {
      const geometries = data.geometries
      return (
        <FeatureGroup {...featureGroupProps}>
          {geometries.map((geometry, i) => (
            <GeoJsonLayer
              {...featureGroupProps}

              key={i}

              data={geometry}
              style={style}
              filter={filter}
              coordsToLatLng={coordsToLatLng}
              eventHandlers={undefined}
              children={undefined}

              eventHandlerFn={eventHandlerFn}
              getChildren={getChildren}
              feature={feature}
              pointAsMarker={pointAsMarker}
              pointProps={pointProps}
              lineProps={lineProps}
              polygonProps={polygonProps}
            />
          ))}
          {children}
        </FeatureGroup>
      )
    }
  }
}

export default React.memo(GeoJsonLayer)
