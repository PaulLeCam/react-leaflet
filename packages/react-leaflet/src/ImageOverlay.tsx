import {
  type MediaOverlayProps,
  createElementObject,
  createLayerComponent,
  extendContext,
  updateMediaOverlay,
} from '@react-leaflet/core'
import { LatLngBounds, ImageOverlay as LeafletImageOverlay } from 'leaflet'
import type { ReactNode } from 'react'

export interface ImageOverlayProps extends MediaOverlayProps {
  children?: ReactNode
  url: string
}
export const ImageOverlay = createLayerComponent<
  LeafletImageOverlay,
  ImageOverlayProps
>(
  function createImageOverlay({ bounds, url, ...options }, ctx) {
    const overlay = new LeafletImageOverlay(url, bounds, options)
    return createElementObject(
      overlay,
      extendContext(ctx, { overlayContainer: overlay }),
    )
  },
  function updateImageOverlay(overlay, props, prevProps) {
    updateMediaOverlay(overlay, props, prevProps)
    if (props.bounds !== prevProps.bounds) {
      const bounds =
        props.bounds instanceof LatLngBounds
          ? props.bounds
          : new LatLngBounds(props.bounds)
      overlay.setBounds(bounds)
    }
    if (props.url !== prevProps.url) {
      overlay.setUrl(props.url)
    }
  },
)
