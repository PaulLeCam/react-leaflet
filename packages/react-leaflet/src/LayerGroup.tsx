import {
  createElementObject,
  createLayerComponent,
  type EventedProps,
  extendContext,
} from '@react-leaflet/core'
import { type LayerOptions, LayerGroup as LeafletLayerGroup } from 'leaflet'
import type { ReactNode } from 'react'

export interface LayerGroupProps extends LayerOptions, EventedProps {
  children?: ReactNode
}

export const LayerGroup = createLayerComponent<
  LeafletLayerGroup,
  LayerGroupProps
>(function createLayerGroup({ children: _c, ...options }, ctx) {
  const group = new LeafletLayerGroup([], options)
  return createElementObject(
    group,
    extendContext(ctx, { layerContainer: group }),
  )
})
