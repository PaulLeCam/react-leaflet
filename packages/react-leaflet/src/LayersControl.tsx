import {
  LeafletProvider,
  createContainerComponent,
  createControlHook,
  createElementHook,
  createElementObject,
  extendContext,
  useLeafletContext,
} from '@react-leaflet/core'
import { Control, type Layer } from 'leaflet'
import React, {
  type ForwardRefExoticComponent,
  type FunctionComponent,
  type ReactNode,
  type RefAttributes,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react'

export interface LayersControlProps extends Control.LayersOptions {
  children?: ReactNode
}

type CustomLayersControlClassConstructor<T extends Control.Layers> = new (baseLayers?: Control.LayersObject, overlays?: Control.LayersObject, options?: Control.LayersOptions) => T

export const useLayersControlElement = <T extends Control.Layers>(
  klass: CustomLayersControlClassConstructor<T>
) => createElementHook<T, LayersControlProps>(
  function createLayersControl({ children: _c, ...options }, ctx) {
    const control = new klass(undefined, undefined, options)
    return createElementObject(
      control,
      extendContext(ctx, { layersControl: control }),
    )
  },
  function updateLayersControl(control, props, prevProps) {
    if (props.collapsed !== prevProps.collapsed) {
      if (props.collapsed === true) {
        control.collapse()
      } else {
        control.expand()
      }
    }
  },
)

export const useLayersControl = <T extends Control.Layers>(klass: CustomLayersControlClassConstructor<T>) => createControlHook(useLayersControlElement(klass))

export interface ControlledLayerProps {
  checked?: boolean
  children: ReactNode
  name: string
}

export function CustomizableLayersControl<T extends Control.Layers>(
  klass: CustomLayersControlClassConstructor<T>
): ForwardRefExoticComponent<
  LayersControlProps & RefAttributes<T>
> & {
  BaseLayer: FunctionComponent<ControlledLayerProps>
  Overlay: FunctionComponent<ControlledLayerProps>
} {
  return Object.assign(createContainerComponent(useLayersControl(klass)), {
    BaseLayer: createControlledLayer(function addBaseLayer(
      layersControl: Control.Layers,
      layer: Layer,
      name: string,
    ) {
      layersControl.addBaseLayer(layer, name)
    }),

    Overlay: createControlledLayer(function addOverlay(
      layersControl: Control.Layers,
      layer: Layer,
      name: string,
    ) {
      layersControl.addOverlay(layer, name)
    }),
  })
}

export const LayersControl = CustomizableLayersControl(Control.Layers)

type AddLayerFunc = (
  layersControl: Control.Layers,
  layer: Layer,
  name: string,
) => void

export function createControlledLayer(addLayerToControl: AddLayerFunc) {
  return function ControlledLayer(props: ControlledLayerProps) {
    const parentContext = useLeafletContext()
    const propsRef = useRef<ControlledLayerProps>(props)
    const [layer, setLayer] = useState<Layer | null>(null)

    const { layersControl, map } = parentContext
    const addLayer = useCallback(
      (layerToAdd: Layer) => {
        if (layersControl != null) {
          if (propsRef.current.checked) {
            map.addLayer(layerToAdd)
          }
          addLayerToControl(layersControl, layerToAdd, propsRef.current.name)
          setLayer(layerToAdd)
        }
      },
      [layersControl, map],
    )
    const removeLayer = useCallback(
      (layerToRemove: Layer) => {
        layersControl?.removeLayer(layerToRemove)
        setLayer(null)
      },
      [layersControl],
    )
    const context = useMemo(() => {
      return extendContext(parentContext, {
        layerContainer: { addLayer, removeLayer },
      })
    }, [parentContext, addLayer, removeLayer])

    useEffect(() => {
      if (layer !== null && propsRef.current !== props) {
        if (
          props.checked === true &&
          (propsRef.current.checked == null ||
            propsRef.current.checked === false)
        ) {
          map.addLayer(layer)
        } else if (
          propsRef.current.checked === true &&
          (props.checked == null || props.checked === false)
        ) {
          map.removeLayer(layer)
        }
        propsRef.current = props
      }
    })

    return props.children ? (
      <LeafletProvider value={context}>{props.children}</LeafletProvider>
    ) : null
  }
}
