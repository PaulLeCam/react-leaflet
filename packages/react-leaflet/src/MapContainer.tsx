import { CONTEXT_VERSION, LeafletProvider } from '@react-leaflet/core'
import {
  FitBoundsOptions,
  LatLngBoundsExpression,
  Map as LeafletMap,
  MapOptions,
} from 'leaflet'
import React, {
  CSSProperties,
  MutableRefObject,
  ReactNode,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react'

export interface MapContainerProps extends MapOptions {
  bounds?: LatLngBoundsExpression
  boundsOptions?: FitBoundsOptions
  children?: ReactNode
  className?: string
  id?: string
  placeholder?: ReactNode
  style?: CSSProperties
  whenCreated?: (map: LeafletMap) => void
  whenReady?: () => void
}

export function useMapElement(
  mapRef: MutableRefObject<HTMLElement | null>,
  props: MapContainerProps,
): LeafletMap | null {
  const [map, setMap] = useState<LeafletMap | null>(null)

  useEffect(() => {
    if (mapRef.current !== null && map === null) {
      const instance = new LeafletMap(mapRef.current, props)
      if (props.center != null && props.zoom != null) {
        instance.setView(props.center, props.zoom)
      } else if (props.bounds != null) {
        instance.fitBounds(props.bounds, props.boundsOptions)
      }
      if (props.whenReady != null) {
        instance.whenReady(props.whenReady)
      }
      setMap(instance)
    }
  }, [mapRef, map, props])

  return map
}

export function MapContainer({
  children,
  className,
  id,
  placeholder,
  style,
  whenCreated,
  ...options
}: MapContainerProps) {
  const mapRef = useRef<HTMLDivElement>(null)
  const map = useMapElement(mapRef, options)

  const createdRef = useRef<boolean>(false)
  useEffect(() => {
    if (map != null && createdRef.current === false && whenCreated != null) {
      createdRef.current = true
      whenCreated(map)
    }
  }, [map, whenCreated])

  const [props] = useState({ className, id, style })
  const context = useMemo(
    () => (map ? { __version: CONTEXT_VERSION, map } : null),
    [map],
  )

  const contents = context ? (
    <LeafletProvider value={context}>{children}</LeafletProvider>
  ) : (
    placeholder ?? null
  )
  return (
    <div {...props} ref={mapRef}>
      {contents}
    </div>
  )
}
