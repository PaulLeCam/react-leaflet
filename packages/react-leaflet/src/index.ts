export { useMap, useMapEvent, useMapEvents } from './hooks.js'

export {
  AttributionControl,
  type AttributionControlProps,
} from './AttributionControl.js'
export { Circle, type CircleProps } from './Circle.js'
export { CircleMarker, type CircleMarkerProps } from './CircleMarker.js'
export { FeatureGroup, type FeatureGroupProps } from './FeatureGroup.js'
export { GeoJSON, type GeoJSONProps } from './GeoJSON.js'
export { ImageOverlay, type ImageOverlayProps } from './ImageOverlay.js'
export { LayerGroup, type LayerGroupProps } from './LayerGroup.js'
export {
  LayersControl,
  CustomizableLayersControl,
  type LayersControlProps,
  type ControlledLayerProps,
} from './LayersControl.js'
export { MapContainer, type MapContainerProps } from './MapContainer.js'
export { Marker, type MarkerProps } from './Marker.js'
export { Pane, type PaneProps } from './Pane.js'
export { Polygon, type PolygonProps } from './Polygon.js'
export { Polyline, type PolylineProps } from './Polyline.js'
export { Popup, type PopupProps } from './Popup.js'
export { Rectangle, type RectangleProps } from './Rectangle.js'
export { ScaleControl, type ScaleControlProps } from './ScaleControl.js'
export { SVGOverlay, type SVGOverlayProps } from './SVGOverlay.js'
export { TileLayer, type TileLayerProps } from './TileLayer.js'
export { Tooltip, type TooltipProps } from './Tooltip.js'
export { VideoOverlay, type VideoOverlayProps } from './VideoOverlay.js'
export { WMSTileLayer, type WMSTileLayerProps } from './WMSTileLayer.js'
export { ZoomControl, type ZoomControlProps } from './ZoomControl.js'
