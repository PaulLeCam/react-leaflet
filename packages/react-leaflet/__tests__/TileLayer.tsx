import { render } from '@testing-library/react'
import React from 'react'

import { MapContainer, TileLayer } from '../src'

describe('TileLayer', () => {
  test('renders', () => {
    const { container, rerender } = render(
      <MapContainer center={[0, 0]} zoom={10}>
        <TileLayer attribution="tiles attribution" url="http://localhost" />
      </MapContainer>,
    )
    expect(container).toMatchSnapshot()

    rerender(
      <MapContainer center={[0, 0]} zoom={10}>
        <TileLayer
          attribution="tiles attribution"
          url="http://localhost?with_parameter=changed"
        />
      </MapContainer>,
    )
    expect(container).toMatchSnapshot()
  })
})
