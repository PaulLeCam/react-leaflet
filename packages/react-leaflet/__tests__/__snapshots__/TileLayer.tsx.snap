// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileLayer renders 1`] = `
<div>
  <div
    class="leaflet-container leaflet-touch leaflet-grab leaflet-touch-drag leaflet-touch-zoom"
    style="position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      >
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 18; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              src="http://localhost"
              style="width: 256px; height: 256px; left: -256px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              src="http://localhost"
              style="width: 256px; height: 256px; left: 0px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              src="http://localhost"
              style="width: 256px; height: 256px; left: -256px; top: 0px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              src="http://localhost"
              style="width: 256px; height: 256px; left: 0px; top: 0px;"
            />
          </div>
        </div>
      </div>
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-disabled="false"
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            <span
              aria-hidden="true"
            >
              +
            </span>
          </a>
          <a
            aria-disabled="false"
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            <span
              aria-hidden="true"
            >
              −
            </span>
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JavaScript library for interactive maps"
          >
            Leaflet
          </a>
           
          <span
            aria-hidden="true"
          >
            |
          </span>
           tiles attribution
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TileLayer renders 2`] = `
<div>
  <div
    class="leaflet-container leaflet-touch leaflet-grab leaflet-touch-drag leaflet-touch-zoom"
    style="position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      >
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 18; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              src="http://localhost?with_parameter=changed"
              style="width: 256px; height: 256px; left: -256px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              src="http://localhost?with_parameter=changed"
              style="width: 256px; height: 256px; left: 0px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              src="http://localhost?with_parameter=changed"
              style="width: 256px; height: 256px; left: -256px; top: 0px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              src="http://localhost?with_parameter=changed"
              style="width: 256px; height: 256px; left: 0px; top: 0px;"
            />
          </div>
        </div>
      </div>
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-disabled="false"
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            <span
              aria-hidden="true"
            >
              +
            </span>
          </a>
          <a
            aria-disabled="false"
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            <span
              aria-hidden="true"
            >
              −
            </span>
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JavaScript library for interactive maps"
          >
            Leaflet
          </a>
           
          <span
            aria-hidden="true"
          >
            |
          </span>
           tiles attribution
        </div>
      </div>
    </div>
  </div>
</div>
`;
